src/Mvc/Mvc.Core/src/ModelBinding/Binders/CollectionModelBinder.cs

```diff
@@ -276,20 +276,15 @@ protected object CreateInstance(Type targetType)
 
         foreach (var value in values)
         {
-            bindingContext.ValueProvider = new CompositeValueProvider
-                {
-                    // our temporary provider goes at the front of the list
-                    new ElementalValueProvider(bindingContext.ModelName, value, values.Culture),
-                    bindingContext.ValueProvider
-                };
-
             // Enter new scope to change ModelMetadata and isolate element binding operations.
             using (bindingContext.EnterNestedScope(
                 elementMetadata,
                 fieldName: bindingContext.FieldName,
                 modelName: bindingContext.ModelName,
                 model: null))
             {
+                bindingContext.ValueProvider = new ElementalValueProvider(bindingContext.ModelName, value, values.Culture);
```

Comment:
@dougbu Do you have any concerns about this change?

Reply:
I'm concerned about everything happening in this PR. It seems very simple to regress behaviours customers are depending on. The existing code looks intentional (though I'm not sure what scenarios could break)&hellip;

I'm also concerned we skip `null` values at all. The problem is likely the `IsModelSet` check just below. May make sense to avoid the new value provider and instead bind `null` when the model type allows that in the context of this loop.

Put another way `["a", "b", null, ""]` and `["a", "b", ""]` should not be treated the same.

> since each element call to the value provider will have the exact same `ModelName`

This basically means the `ElementValueProvider` will always match and provide a result; `null` is a value `Value` in many contexts. So, what would this change do anyway❔ How would using the `CompositeValueProvider` cause duplicates in the resulting array❔

Reply:
> How would using the CompositeValueProvider cause duplicates in the resulting array❔

Because this code overrides and reuses the `bindingContext.ValueProvider`:

```c#
foreach (var value in values)
{
    // bindingContext.ValueProvider is overridden
    bindingContext.ValueProvider = new CompositeValueProvider
        {
            new ElementalValueProvider(bindingContext.ModelName, value, values.Culture),
            bindingContext.ValueProvider // this uses the value provider from the previous iteration
        };

    ...
}
```

A change which would just fix the "value from previous item/iteration" could look like this:

```c#
var initalValueProvider = bindingContext.ValueProvider;
foreach (var value in values)
{
    // bindingContext.ValueProvider is overridden
    bindingContext.ValueProvider = new CompositeValueProvider
        {
            new ElementalValueProvider(bindingContext.ModelName, value, values.Culture),
            initalValueProvider
        };

    ...
}
```

See also https://github.com/dotnet/aspnetcore/pull/42278#discussion_r901924127

Reply:
Ah, I finally get it 😃 I like that approach much better. The duplicate values are the result of a straight-up bug and nobody should rely on that behaviour.

I'm not sure about the dropped `null` values. They're obviously due to the `IsModelSet` check just below. But fixing that issue feels like a behavioural break, And special cases for many item types will complicate things.

Could we fix the duplication then see how customers react to `["a", "b", null, null, "c"]` becoming `["a", "b", "c"]` or `["a", "b", "", "", "c"]` for a `string[]`, depending on the `ConvertEmptyStringToNull` setting❔

Reply:
I have adjusted the code to keep the `CompositeValueProvider`.